### DISCUSSION TITLE
[PUI] Select component with async support

### DISCUSSION THREAD

**Discussion by SchrodingersGat** [inventree/InvenTree#5495] [General]
I am currently implementing modal forms / API forms in https://github.com/inventree/InvenTree/pull/5355. This is a major refactor of the existing forms interface, rewritten for react / mantine.

However, I am running into a blocking issue - async data fetching support using the [Select](https://mantine.dev/core/select/) component provided by mantine.

By default, the select component does not support async data fetching. However, it can be implemented using various hooks. Some information is available here: 

- https://github.com/orgs/mantinedev/discussions/345
- https://github.com/orgs/mantinedev/discussions/4077

I have made some progress here, but the implementation is not anywhere near as good as the existing javascript implementation which uses [select2](https://select2.org)

In particular, it appears quite difficult to support "infinite pagination" - that is, fetching new results as the user scrolls down through the available options. In fact in [this comment](https://github.com/orgs/mantinedev/discussions/345#discussioncomment-4280191) the developer states it is not supported at all.

Good async support for these dropdown is critical for our API / modal form architecture. 

@matmair @wolflu05 any thoughts on this one?

Perhaps we should integrate [react select](https://react-select.com/home)? How would this "theme" in with mantine? Can we mix and match components like this?

**Comment by SchrodingersGat** [comment]
An alternative option (if this is even possible) might be to integrate something like [react-window-infinite-loader](https://github.com/bvaughn/react-window-infinite-loader) instead of the native mantine select dropdown?

**Comment by wolflu05** [comment]
With that information and support from mantine, I would suggest using the react select library with the [`onMenuScrollToBottom`](https://react-select.com/props#select-props) prop. This could be integrated into the matine theming using the [styles api](https://mantine.dev/styles/styles-api/) and maybe also [custom components](https://mantine.dev/guides/custom-components/) are helpful. React hooks are really powerful and if you combine them all you can build something really cool.

**Reply by matmair** [reply]
That should work - although I would hold off on styling till we have a final design

**Reply by SchrodingersGat** [reply]
Thanks for the input @wolflu05 I'll look at roughing in some functionality

**Comment by matmair** [comment]
@SchrodingersGat [this](https://github.com/orgs/mantinedev/discussions/4077#discussioncomment-6702299) looks like a good solution that could probably be used generically.

**Reply by SchrodingersGat** [reply]
@matmair yes I looked at this yesterday, but (as far as I can tell) it is not possible to implement an infinite pagination? Will investigate further
